mmedit/models/restorers/tdan.py

```diff
@@ -0,0 +1,143 @@
+        pretrained (str): Path for pretrained model. Default: None.
+    """
+
+    def __init__(self,
+                 generator,
+                 pixel_loss,
+                 lr_pixel_loss,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        super().__init__(generator, pixel_loss, train_cfg, test_cfg,
+                         pretrained)
+
+        self.lr_pixel_loss = build_loss(lr_pixel_loss)
+
+    def forward_train(self, lq, gt):
+        """Training forward function.
+
+        Args:
+            lq (Tensor): LQ Tensor with shape (n, c, h, w).
+            gt (Tensor): GT Tensor with shape (n, c, h, w).
+
+        Returns:
+            Dict: Output dictionary containing necessary information.
+        """
+        losses = dict()
+        output, aligned_lrs = self.generator(lq)
+
+        # loss on the HR image
+        loss_pix = self.pixel_loss(output, gt)
+        losses['loss_pix'] = loss_pix
+        # loss on the aligned LR images
+        t = aligned_lrs.size(1)
+        lr_ref = lq[:, t // 2].unsqueeze(1).repeat(1, t, 1, 1, 1)
+        loss_pix_lr = self.lr_pixel_loss(aligned_lrs, lr_ref)
+        losses['loss_pix_lr'] = loss_pix_lr
+
+        outputs = dict(
+            losses=losses,
+            num_samples=len(gt.data),
+            results=dict(lq=lq.cpu(), gt=gt.cpu(), output=output.cpu()))
+
+        return outputs
+
+    def evaluate(self, output, gt):
+        """Evaluation function.
+
+        Args:
+            output (Tensor): Model output with shape (n, c, h, w).
+            gt (Tensor): GT Tensor with shape (n, c, h, w).
+
+        Returns:
+            dict: Evaluation results.
+        """
+        crop_border = self.test_cfg.crop_border
+        convert_to = self.test_cfg.get('convert_to', None)
+
+        output = tensor2img(output)
+        gt = tensor2img(gt)
+
+        eval_result = dict()
+        for metric in self.test_cfg.metrics:
+            eval_result[metric] = self.allowed_metrics[metric](
+                output, gt, crop_border, convert_to=convert_to)
+        return eval_result
+
+    def forward_test(self,
+                     lq,
+                     gt=None,
+                     meta=None,
+                     save_image=False,
+                     save_path=None,
+                     iteration=None):
+        """Testing forward function.
+
+        Args:
+            lq (Tensor): LQ Tensor with shape (n, c, h, w).
+            gt (Tensor): GT Tensor with shape (n, c, h, w). Default: None.
+            save_image (bool): Whether to save image. Default: False.
+            save_path (str): Path to save image. Default: None.
+            iteration (int): Iteration for the saving image name.
+                Default: None.
+
+        Returns:
+            dict: Output results.
+        """
+        output = self.generator(lq)[0]  # keep only the HR output
+
+        # normalize from [-0.5, 0.5] to [0, 1] (following official)
```

Comment:
We may unify the processing of input/output normalization/de-normalization next month. This is hard-coded everywhere